File: FindMin.js
class FindMin {
    public static int findMin(int arr[],int low,int high) 
  { 
    
      if (low==high) return arr[low]; 

      int mid = low + (high - low)/2; 

      if (mid < high && arr[mid+1] < arr[mid]) 
          return arr[mid+1]; 

      if (mid > low && arr[mid] < arr[mid - 1]) 
          return arr[mid]; 

      if (arr[high] > arr[mid]) {
          return findMin(arr, low, mid-1); 
      }
      return findMin(arr, mid+1, high);
  }  
  
  public static void main (String[] args) {
      int nums[]={1,2,3,4};
      int min =findMin(nums,0,nums.length -1);
      System.out.println(min);
      
  }
}

File: findMin.cc
class Solution {
public:
    int findMin(vector<int>& nums) {
        int low = 0;
        int high = nums.size() - 1;
        while(low <= high){
            //base case: if the array turns out to be sorted
            if(nums[low] < nums[high]){
                return nums[low];
            }
            int mid = low + (high - low)/2;
            //The minimum element is always lesser than its immediate neighbbors
            if((mid == low || nums[mid] < nums[mid - 1]) && (mid == high || nums[mid] < nums[mid + 1])){
                return nums[mid];
            }
            
            if(nums[low] <= nums[mid]){
                low = mid + 1;
            }
            else{
                high = mid - 1;
            }
        }
        return -1;
    }
};